src/typing_extensions.py

```diff
@@ -829,35 +853,73 @@ def __new__(cls, name, bases, ns, total=True):
                 }
             required_keys = set()
             optional_keys = set()
+            readonly_keys = set()
+            mutable_keys = set()
 
             for base in bases:
                 annotations.update(base.__dict__.get('__annotations__', {}))
                 required_keys.update(base.__dict__.get('__required_keys__', ()))
                 optional_keys.update(base.__dict__.get('__optional_keys__', ()))
+                readonly_keys.update(base.__dict__.get('__readonly_keys__', ()))
+                mutable_keys.update(base.__dict__.get('__mutable_keys__', ()))
+
+                if not getattr(base, "__other_keys__", True):
+                    if (
+                        set(own_annotations) -
+                        set(base.__dict__.get('__annotations__', {}))
+                    ):
+                        raise TypeError(
+                            "TypedDict cannot inherit from a TypedDict with "
+                            "other_keys=False and new fields"
+                        )
+                if (
+                    readonly
+                    and is_typeddict(base)
+                    and base is not _TypedDict
+                    and not getattr(base, "__readonly__", False)
+                ):
+                    raise TypeError(
+                        "read-only TypedDict cannot extend non-read-only "
+                        f"TypedDict {base}"
+                    )
 
             annotations.update(own_annotations)
             for annotation_key, annotation_type in own_annotations.items():
-                annotation_origin = get_origin(annotation_type)
-                if annotation_origin is Annotated:
-                    annotation_args = get_args(annotation_type)
-                    if annotation_args:
-                        annotation_type = annotation_args[0]
-                        annotation_origin = get_origin(annotation_type)
-
-                if annotation_origin is Required:
+                qualifiers = list(_get_typeddict_qualifiers(annotation_type))
```

Comment:
Very minor, but a `set` might be better here (it would make the `if Required in qualifiers` check below O(1) )

```suggestion
                qualifiers = set(_get_typeddict_qualifiers(annotation_type))
```

Reply:
Done